internal/testplanet/planet.go

```diff
@@ -323,6 +343,106 @@ func (planet *Planet) Shutdown() error {
 	return errlist.Err()
 }
 
+// Ping sends a kad ping request to/from all kad nodes to each other
+// NB: kad nodes include bootstrap, satellite, and storagenode type nodes
+func (planet *Planet) Ping(ctx *testcontext.Context) error {
+	pingErrs := permutePeers(planet, func(a, b Peer, labels [2]string) error {
+		if a == b {
+			return nil
+		}
+		return kadPing(ctx, a, b, labels)
+	})
+	return pingErrs
+}
+
+// permutePeers calls the passed function for each permutation of peers
+// (except for a peer with itself, also except uplink permutations, presently)
+func permutePeers(planet *Planet, f permutePeersFunc) error {
+	return iteratePeers(planet, func(a Peer, labelA string) error {
+		return iteratePeers(planet, func(b Peer, labelB string) error {
+			if a == b {
+				return nil
+			}
+			return f(a, b, [2]string{labelA, labelB})
+		})
+	})
+}
+
+// iterate peers calls the passed function for each peer in the planet
+// (except uplink, presently)
+func iteratePeers(planet *Planet, f iteratePeersFunc) error {
+	errGroup := errs.Group{}
+
+	errGroup.Add(f(planet.Bootstrap, "bootstrap"))
+
+	for i, s := range planet.Satellites {
+		errGroup.Add(f(s, fmt.Sprintf("satellite %d", i)))
+	}
+
+	for i, s := range planet.StorageNodes {
+		errGroup.Add(f(s, fmt.Sprintf("storagenode %d", i)))
+	}
+
+	// uplink
+	// TODO: do something here?
+	return errGroup.Err()
+}
+
+func kadPing(ctx *testcontext.Context, local, remote Peer, labels [2]string) (err error) {
+	switch p := local.(type) {
+	case *bootstrap.Peer:
+		_, err = p.Kademlia.Service.Ping(ctx, remote.Local())
+	case *satellite.Peer:
+		_, err = p.Kademlia.Service.Ping(ctx, remote.Local())
+	case *storagenode.Peer:
+		_, err = p.Kademlia.Service.Ping(ctx, remote.Local())
+	default:
+		return errs.New("unknown peer type: %T", p)
+	}
+	return kadPingErr(local, remote, labels, err)
+}
+
+func kadPingErr(local, remote Peer, labels [2]string, err error) error {
+	if err == nil {
+		return nil
+	}
+	return ErrPing.New("%s (%.5s) --> %s (%.5s): %s", labels[0], local.ID(), labels[1], remote.ID(), err.Error())
+}
+
+func writeWhitelist(dir string) (string, error) {
+	whitelistPath := filepath.Join(dir, "whitelist.pem")
+	signer := NewPregeneratedSigner()
+	err := identity.PeerCAConfig{
+		CertPath: whitelistPath,
+	}.Save(signer.PeerCA())
+
+	return whitelistPath, err
+}
+
+// usePeerCAWhitelist reconfigures the planet to use peer ca whitelisting on all kad node types
+func (planet *Planet) usePeerCAWhitelist() (Reconfigure, error) {
+	whitelistPath, err := writeWhitelist(planet.directory)
+	if err != nil {
+		return Reconfigure{}, err
+	}
+
+	return Reconfigure{
+		Bootstrap: func(i int, c *bootstrap.Config) {
+			c.Server.PeerCAWhitelistPath = whitelistPath
```

Comment:
Maybe always include the `PeerCAWhitelistPath` and then just toggle the `UseCAWhitelist` boolean in the initial setup.

Reply:
i was mistakenly under the impression that if `PeerCAWhitelist` was specified then `UseCAWhitelist` was turned on but I now see that's not the case; good suggestion.